components/director/internal/authenticator/middleware.go

```diff
@@ -8,57 +8,81 @@ import (
 	"strings"
 	"sync"
 
-	"github.com/lestrrat-go/iter/arrayiter"
-
-	"github.com/kyma-incubator/compass/components/director/pkg/authenticator"
-
+	"github.com/form3tech-oss/jwt-go"
+	"github.com/kyma-incubator/compass/components/director/internal/authenticator/claims"
 	"github.com/kyma-incubator/compass/components/director/internal/domain/client"
-
 	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
-
-	"github.com/kyma-incubator/compass/components/director/internal/consumer"
-
+	"github.com/kyma-incubator/compass/components/director/pkg/authenticator"
 	"github.com/kyma-incubator/compass/components/director/pkg/log"
-	"github.com/pkg/errors"
-
-	"github.com/kyma-incubator/compass/components/director/internal/domain/tenant"
-	"github.com/kyma-incubator/compass/components/director/pkg/scope"
-
-	"github.com/form3tech-oss/jwt-go"
+	"github.com/lestrrat-go/iter/arrayiter"
 	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/pkg/errors"
 )
 
 const (
 	AuthorizationHeaderKey = "Authorization"
 	JwksKeyIDKey           = "kid"
 )
 
+//go:generate mockery --name=ClaimsParser --output=automock --outpkg=automock --case=underscore
+type ClaimsParser interface {
+	ParseClaims(ctx context.Context, bearerToken string, keyfunc jwt.Keyfunc) (claims.Claims, error)
+}
+
 type Authenticator struct {
-	jwksEndpoint        string
+	jwksEndpoints       []string
 	allowJWTSigningNone bool
 	cachedJWKS          jwk.Set
 	clientIDHeaderKey   string
 	mux                 sync.RWMutex
+	claimsParser        ClaimsParser
 }
 
-func New(jwksEndpoint string, allowJWTSigningNone bool, clientIDHeaderKey string) *Authenticator {
+// TODO clientIDHeaderKey
+func New(allowJWTSigningNone bool, clientIDHeaderKey string, claimsParser ClaimsParser, jwksEndpoints ...string) *Authenticator {
 	return &Authenticator{
-		jwksEndpoint:        jwksEndpoint,
+		jwksEndpoints:       jwksEndpoints,
 		allowJWTSigningNone: allowJWTSigningNone,
 		clientIDHeaderKey:   clientIDHeaderKey,
+		claimsParser:        claimsParser,
 	}
 }
 
 func (a *Authenticator) SynchronizeJWKS(ctx context.Context) error {
 	log.C(ctx).Info("Synchronizing JWKS...")
 	a.mux.Lock()
 	defer a.mux.Unlock()
-	jwks, err := authenticator.FetchJWK(ctx, a.jwksEndpoint)
-	if err != nil {
-		return errors.Wrapf(err, "while fetching JWKS from endpoint %s", a.jwksEndpoint)
+	newJWKS := jwk.NewSet()
+
+	for _, jwksEndpoint := range a.jwksEndpoints {
+		jwks, err := authenticator.FetchJWK(ctx, jwksEndpoint)
+		if err != nil {
+			return errors.Wrapf(err, "while fetching JWKS from endpoint %s", jwksEndpoint)
+		}
+		keyIterator := &authenticator.JWTKeyIterator{
+			AlgorithmCriteria: func(alg string) bool {
+				return true
+			},
+			IDCriteria: func(id string) bool {
+				return true
+			},
+		}
+
+		if err := arrayiter.Walk(ctx, jwks, keyIterator); err != nil {
+			return errors.Wrapf(err, "while walking through JWKS")
+		}
+
+		for _, key := range keyIterator.AllKeys {
+			key, ok := key.(jwk.Key)
```

Comment:
The type assertion is already made by the keyIterator. We can replace the loop body with
```code
newJWKS.Add(key.(jwk.Key))
```